beacon-chain/sync/validate_sync_committee_message.go

```diff
@@ -0,0 +1,278 @@
+) pubsub.ValidationResult {
+	ctx, span := trace.StartSpan(ctx, "sync.validateSyncCommitteeMessage")
+	defer span.End()
+
+	if pid == s.cfg.P2P.PeerID() {
+		return pubsub.ValidationAccept
+	}
+
+	// Basic validations before proceeding.
+	if s.cfg.InitialSync.Syncing() {
+		return pubsub.ValidationIgnore
+	}
+	if msg.Topic == nil {
+		return pubsub.ValidationReject
+	}
+
+	// Read the data from the pubsub message, and reject if there is an error.
+	m, err := s.readSyncCommitteeMessage(msg)
+	if err != nil {
+		log.WithError(err).Debug("Could not decode message")
+		traceutil.AnnotateError(span, err)
+		return pubsub.ValidationReject
+	}
+
+	// Validate sync message times before proceeding.
+	// The message's `slot` is for the current slot (with a MAXIMUM_GOSSIP_CLOCK_DISPARITY allowance).
+	if err := altair.ValidateSyncMessageTime(
+		m.Slot,
+		s.cfg.Chain.GenesisTime(),
+		params.BeaconNetworkConfig().MaximumGossipClockDisparity,
+	); err != nil {
+		traceutil.AnnotateError(span, err)
+		return pubsub.ValidationIgnore
+	}
+
+	committeeIndices, err := s.cfg.Chain.HeadCurrentSyncCommitteeIndices(ctx, m.ValidatorIndex, m.Slot)
+	if err != nil {
+		traceutil.AnnotateError(span, err)
+		return pubsub.ValidationIgnore
+	}
+
+	// Validate the message's data according to the p2p specification.
+	if result := validationPipeline(
+		ctx,
+		ignoreEmptyCommittee(committeeIndices),
+		s.rejectIncorrectCommittee(committeeIndices, *msg.Topic),
+		s.ignoreHasSeenSyncMsg(m, committeeIndices),
+		s.rejectInvalidSignature(m),
+	); result != pubsub.ValidationAccept {
+		return result
+	}
+
+	s.markSyncCommitteeMessagesSeen(committeeIndices, m)
+
+	msg.ValidatorData = m
+	return pubsub.ValidationAccept
+}
+
+// Parse a sync committee message from a pubsub message.
+func (s *Service) readSyncCommitteeMessage(msg *pubsub.Message) (*ethpb.SyncCommitteeMessage, error) {
+	raw, err := s.decodePubsubMessage(msg)
+	if err != nil {
+		return nil, err
+	}
+	m, ok := raw.(*ethpb.SyncCommitteeMessage)
+	if !ok {
+		return nil, errWrongMessage
+	}
+	if m == nil {
+		return nil, errNilMessage
+	}
+	return m, nil
+}
+
+// Mark all a slot and validator index as seen for every index in a committee and subnet.
+func (s *Service) markSyncCommitteeMessagesSeen(committeeIndices []types.CommitteeIndex, m *ethpb.SyncCommitteeMessage) {
+	subCommitteeSize := params.BeaconConfig().SyncCommitteeSize / params.BeaconConfig().SyncCommitteeSubnetCount
+	for _, idx := range committeeIndices {
+		subnet := uint64(idx) / subCommitteeSize
+		s.setSeenSyncMessageIndexSlot(m.Slot, m.ValidatorIndex, subnet)
+	}
+}
+
+// Returns true if the node has received sync committee for the validator with index and slot.
+func (s *Service) hasSeenSyncMessageIndexSlot(slot types.Slot, valIndex types.ValidatorIndex, subCommitteeIndex uint64) bool {
+	s.seenSyncMessageLock.RLock()
+	defer s.seenSyncMessageLock.RUnlock()
+
+	b := append(bytesutil.Bytes32(uint64(slot)), bytesutil.Bytes32(uint64(valIndex))...)
```

Comment:
Can refactor the following into a key function to share between `hasSeen` and `setSeen`
```go
b := append(bytesutil.Bytes32(uint64(slot)), bytesutil.Bytes32(uint64(valIndex))...)
b = append(b, bytesutil.Bytes32(subCommitteeIndex)...)
```